Multitenant payment processor config

e-kolpakov opened the pull request:
**Description:** Adds multitenancy support for payment processor configuration, using Django sites framework.
**Dependencies:** #629 
**Merge deadline:** 1 Apr
**Related:** [Discovery document]
(https://docs.google.com/document/d/17BRYvkO4ETvBV9TFodElpbJaK4oze7e0mNqf9ztnGnA/edit#heading=h.70eeasm5mr1q)
**Sandbox:** 
* edX site: [LMS](http://pr11898.sandbox.opencraft.com), [Studio](http://studio.pr11898.sandbox.opencraft.com), [eCommerce](https://ecommerce.pr11898.sandbox.opencraft.com)
* WL site: [LMS](http://wlsite.pr11898.sandbox.opencraft.com), [eCommerce](https://ecommerce.wlsite.pr11898.sandbox.opencraft.com)

Note both HTTP and HTTPS are available for LMS and Otto.

**JIRA:** [WL-362](https://openedx.atlassian.net/browse/WL-362)


**Preparing environment**
1. Switch to current branch
2. Modify settings - move current contents of PAYMENT_PROCESSOR_CONFIG into a new nested dict with  `default` key, i.e. 
    `PAYMENT_PROCESSOR_CONFIG = { 'default': OLD_PAYMENT_PROCESSOR_CONFIG_VALUE }`
3. Migrate DB (adds "OAuth settings" field to `SiteConfiguration`) model.

**LMS and Otto configuration**

The idea is to set up two Sites in Otto and configure Otto and LMS so that both sites are operational.
We will be using two "sites": "edx" and "wl"; so [sitename] is one of (`edx`,  `wl`)

**Configuring LMS:**

The goal here is to configure two (or more) OAuth clients.

1. Go to http://localhost:8000/admin/oauth2/client/ -> Add Client
2. Set values: 
  1. *Url:* http://[sitename].lms.local:8002
  2. *Redirect Uri*: http://[sitename].lms.local:8002/complete/edx-oidc/
  3. *Client Id:* ecommerce-[sitename]-key
  4. *Client Secret:* ecommerce-[sitename]-secret
  5. *Client type:* Confidential
3. Go to http://localhost:8000/admin/oauth2_provider/trustedclient/ and designate both clients added in (2) as trusted (add one record per client)
4. In http://localhost:8000/admin/commerce/commerceconfiguration/ add/edit the configuration and set check "Checkout on ecommerce service".
5. Make sure FEATURES['ENABLE_OAUTH2_PROVIDER'] is set to True in lms.env.json (required to generate credit course records in course catalog - can be skipped if no credit-mode courses are used)

**Configuring Otto**

Otto configuration: Configure two Sites and SiteConfiguration records so that Otto can connect to LMS

1. Add two site records at http://localhost:8002/admin/sites/site/ (i.e. edx.lms.local:8002 and wl.lms.local:8002)
2. Add two partner records (http://localhost:8002/admin/partner/partner/) - choose unique name and short_code, don't select users.
3. Add corresponding SiteConfiguration records (http://mitpe.lms.analytics.local:8002/admin/core/siteconfiguration/):
  1. *Site:* choose one of sites selected earlier
  2. *Parner:* choose unique partner (or create new)
  3. *Path to custom site theme:* anything, but empty string (required, but works fine if file does not exist)
  4. *Payment processors:' comma separated list of supported processors (supported so far: paypal, cybersource), i.e. `paypal,cybersource`
  5. OAuth settings:
```
    {
      "SOCIAL_AUTH_EDX_OIDC_ID_TOKEN_DECRYPTION_KEY":"ecommerce-[sitename]-secret",
      "SOCIAL_AUTH_EDX_OIDC_KEY":"ecommerce-[sitename]-key",
      "SOCIAL_AUTH_EDX_OIDC_SECRET":"ecommerce-[sitename]-secret",
      "SOCIAL_AUTH_EDX_OIDC_URL_ROOT": "http://[sitename].lms.local:8000/oauth2",
      "OAUTH_PROVIDER_URL":"http://[sitename].lms.local:8000/"
    }
```

**Set up payment processor configuration in Otto**

Add payment processor configuration for each site - add new top level records to PAYMENT_PROCESSOR_CONFIG setting:

    PAYMENT_PROCESSOR_CONFIG = {
        'default': { ... },
        '[sitename]': { 
            'cybersource': { $cybersource_config$ },
            'paypal': { $paypal_config$ },
    }

$cybersource_config$ simplest way to do everything right is to copy PAYMENT_PROCESSOR_CONFIG['default'] value and than replace relevant settings. For cybersource 'merchant_id', 'transaction_key':  'profile_id', 'access_key',  'secret_key' should be replaced.


**Configure WL microsite in LMS**

TBD - this part is reworked by @mattdrayer (microsite -> comprehensive theming). Meanwhile, just continue using microsite framework - add the following to lms.env.json:
    
```
"MICROSITE_CONFIGURATION": {
        "wl": {
                        "cybersource_config_key": "wl",
                        "domain_prefix": "wl",
                        "university": "wl",
                        "platform_name": "White Label Site",
                        "css_overrides_file": "css/main.css",
                        "SITE_NAME": "wl.exampl.ecom",
                        "course_org_filter": "wlX",
                        "show_partners": false,
                        "ECOMMERCE_PUBLIC_URL_ROOT":  "http://[sitename].lms.local:8002/"
                }
    },
```

LMS does not support checkout with Otto for themed sites/microsites in master branch, so you'd need to switch to [alternative branch](https://github.com/open-craft/edx-platform/tree/ekolpakov/theme-ecommerce-support) until it is merged to master.

**Generate course catalog entries**
1. Create courses in Studio (documented elsewhere)
2. Verify both main site and WL site have some courses - go to LMS (at edx.lms.local and wl.lms.local) and make sure there are some courses displayed in "Find Courses".
3. Go to http://localhost:8002/courses/ and add new course records to course catalog (documented elsewhere)

**Update host machine hosts file**
Add edx.lms.local and wl.lms.local to `/etc/hosts` (or equivalent) as aliases of 127.0.0.1

**Testing instructions:**
For both sites (wl and edx):
1. Log out from both Otto (localhost:8002) and LMS (localhost:8000 / edx.lms.local / wl.lms.local)
2. Log in to LMS as honor (or create new user)
3. Go to course catalogue
4. Enroll in course (if Verified track is avalilable) and upgrade to "Verified" track or add course to a cart (if "professional" track is available).
5. You'll arrive at checkout page. This page contains one "Checkout" button per available payment processor (Cybersource/Paypal). *Expected result:* buttons are shown only for processors enabled in SiteConfiguration model for that site.
6. Click "Checkout" - you'll be taken to the payment processor site for payment. *Expected result:* correct PAYMENT_PROCESSOR_CONFIG entry is used. The simplest would be to design visually different payment pages for different configs (i.e. include different subset of optional fields).
7. Complete payment
8. You should be redirected back to appropriate microsite/main site (all payment processors), and have the course upgraded to Verified track (only Paypal - Cybersource can not connect to VM boxes, unless they are exposed to the world - sandbox to test that is currently provisioning)

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=630&repo=edx%2Fecommerce

jbzdak commented on ecommerce/core/models.py:
```diff
@@ -34,10 +34,25 @@ class SiteConfiguration(models.Model):
 
+    @property
+    def allowed_payment_processors(self):
+        payment_processors = (
```
Possibly (for backward compatibility) it would be good to have empty `payment_processors` mean 'every processor', as this field my be unfilled on various production instances.

e-kolpakov replied:
@jbzdak maybe, but if there are production instance(s) with empty `payment_processor` fields, it would be *very* easy to spot (no "checkout" buttons) and easy to fix. Also, it is a direct violation of setup instructions, so I don't think there are any.

So, in my opinion, we shouldn't add this little workaround/hack to support backward compatibility. Let's see if upstream reviewer (@douglashall ?) have something to say about it.

douglashall replied:
I do not think that an empty `payment_processors` field should convert to all payment processors.

jbzdak commented on ecommerce/core/models.py:
```diff
@@ -34,10 +34,25 @@ class SiteConfiguration(models.Model):
+    def allowed_payment_processors(self):
+        payment_processors = (
+            raw_processor_value.strip()
+            for raw_processor_value in self.payment_processors.split(',')
```
If we are considering maintainability of this code, it would be good to move this to ``EdxOrderPlacementMixin``, as this is the type that requires ``self.payment_processor`` to be a defined instance, and define payment processor type as a class constant.

e-kolpakov replied:
@jbzdak are you sure this is the right place for this comment? If it is, than no, this setting should stay here, as it just transforms comma-separated `payment_processors` field into a set of allowed payment processor names.

Also, only two children of `EdxOrderPlacementMixin` uses `payment_processor` property, so it does not look like `payment_processor` belongs to it.

jbzdak replied:
OK you are right --- i pasted it in the wrong place (I was moving comments from #632 to here), and you are right on that that there are more subclasses ``EdxOrderPlacementMixin`` than I was aware of.

jbzdak commented:
@e-kolpakov It works as advertised, ping me when you finish --- add tests and so on --- and I'll re-check :)

jbzdak commented on ecommerce/extensions/payment/config.py:
```diff
@@ -7,7 +7,7 @@ class PaymentConfig(config.PaymentConfig):
 
     def ready(self):
-        paypal_configuration = settings.PAYMENT_PROCESSOR_CONFIG.get('paypal')
+        paypal_configuration = settings.PAYMENT_PROCESSOR_CONFIG['default'].get('paypal')
```
Please double-check this actually makes sense. ``paypalrestsdk`` could have different config than currently used microsite, which could lead to trouble.

jbzdak commented on ecommerce/settings/devstack.py:
```diff
@@ -23,25 +21,27 @@
-        'cancel_url': get_lms_url('/commerce/checkout/cancel/'),
-        'error_url': get_lms_url('/commerce/checkout/error/'),
-    },
+    'default': {
```
Default is a misleading name here. I read default as: "If I don't override settings for a my microsite, it will use default settings". Outside test scope it is only used [here](https://github.com/edx/ecommerce/pull/630/files#r56902761) (I think), so possibly it should just go away.

jbzdak commented:
@e-kolpakov Thank you for very detailed setup instructions :)

jbzdak commented on ecommerce/extensions/payment/views.py:
```diff
@@ -35,7 +35,9 @@
-    payment_processor = Cybersource()
+    @property
+    def payment_processor(self):
+        return Cybersource()
```
If we are considering maintainability of this code, it would be good to move this to EdxOrderPlacementMixin, as this is the type that requires self.payment_processor to be a defined instance, and define payment processor type as a class constant.

jbzdak commented on ecommerce/core/tests/test_models.py:
```diff
@@ -38,3 +40,21 @@ def test_get_full_name(self):
+class SiteConfigurationTests(TestCase):
+    @ddt.data(
+        ("", set()),
+        (",,", set()),
```
I like this case :)

clintonb replied:
Validate the data before it's stored, and you don't have to worry about cases like these.

jbzdak commented on ecommerce/core/tests/test_models.py:
```diff
@@ -38,3 +40,21 @@ def test_get_full_name(self):
+        ("paypal,", {"paypal"}),
+        ("paypal,cybersource", {"paypal", "cybersource"}),
+        ("paypal,,cybersource", {"paypal", "cybersource"}),
+        (",paypal,,cybersource", {"paypal", "cybersource"}),
```
Maybe add something like `, paypal   , ,  cybersource`, to check for regressions when spaces arn't properly stripped.

jbzdak commented on ecommerce/extensions/payment/helpers.py:
```diff
@@ -87,3 +89,18 @@ def sign(message, secret):
+        available_processors = [
+            processor for processor in all_processors
+            if (
+                processor.NAME in request.site.siteconfiguration.allowed_payment_processors and
```
NIT: Create a local variable for `request.site.siteconfiguration.allowed_payment_processors`, and then whole if would fit in a single line :)

jbzdak commented on ecommerce/extensions/payment/tests/test_helpers.py:
```diff
@@ -40,3 +44,58 @@ def test_sign(self):
+        ([DummyProcessor, AnotherDummyProcessor], [DummyProcessor, AnotherDummyProcessor]),
+    )
+    @ddt.unpack
+    def test_get_payment_processors(self, processors, expected_result):
```
It looks like in every case ``processors==expected_result`` so maybe either change them to vary or merge them to a single variable.

e-kolpakov replied:
@jbzdak that's just a coincidence - expected result have the same format and value as test input, so I would rather keep it separated (for flexibility) and in line with other tests.

jbzdak replied:
OK. Agreed.

jbzdak commented:
Solid code +1, I have verified using provided sandbox. Before the merge we need to verify that there is proper redirection (wlsite.lms redirects to wlsite.otto and plain lms redirects to plain otto).

bradenmacdonald commented:
@e-kolpakov Can you please rebase this too, now that Doug's PR has merged?

e-kolpakov commented:
@clintonb @douglashall please review.

douglashall commented on ecommerce/core/models.py:
```diff
@@ -45,6 +45,15 @@ class SiteConfiguration(models.Model):
         unique_together = ('site', 'partner')
 
+    @property
+    def allowed_payment_processors(self):
```
Can we rename this to `payment_processors_set`?

douglashall commented on ecommerce/core/models.py:
```diff
@@ -45,6 +45,15 @@ class SiteConfiguration(models.Model):
+            for raw_processor_value in self.payment_processors.split(',')
+            if raw_processor_value and not raw_processor_value.isspace()
+        )
+        return set(payment_processors)
```
You could use a set comprehension here instead.

```
return {
    raw_processor_value.strip()
    for raw_processor_value in self.payment_processors.split(',')
    if raw_processor_value and not raw_processor_value.isspace()
}
```

douglashall commented on ecommerce/extensions/payment/helpers.py:
```diff
@@ -87,3 +89,16 @@ def sign(message, secret):
+            processor for processor in all_processors
+            if processor.NAME in site_config.allowed_payment_processors and processor.is_enabled()
+        ]
+        return available_processors
```
You could reduce this to a single statement.
```
request = get_current_request()
if request:
    allowed_payment_processors = request.site.siteconfiguration.allowed_payment_processors
    return [
        processor for processor in all_processors
        if processor.is_enabled() and processor.NAME in allowed_payment_processors
    ]
```

douglashall commented on ecommerce/extensions/payment/management/commands/paypal_profile.py:
```diff
@@ -26,7 +28,7 @@ class Command(BaseCommand):
     The 'enable' and 'disable' actions are idempotent so it is safe to run them repeatedly in the same environment.
     """
-    args = "action [id] [json]"
+    args = "action config [id] [json]"
```
Rename `config` to `partner`.

douglashall commented:
:+1: once comments addressed. Thanks!

douglashall commented:
@e-kolpakov We will also need a corresponding PR in against the configuration [repo](https://github.com/edx/configuration/blob/master/playbooks/roles/ecommerce/defaults/main.yml#L82) to update the payment processor configuration.

e-kolpakov commented:
@douglashall Config repo PR: https://github.com/edx/configuration/pull/2886

e-kolpakov commented:
@clintonb do you want to review this, or is @douglashall approval enough for the merge?

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +45,14 @@ class SiteConfiguration(models.Model):
+        return {
+            raw_processor_value.strip()
+            for raw_processor_value in self.payment_processors.split(',')
+            if raw_processor_value and not raw_processor_value.isspace()
```
Validate the data before it's stored, and you don't have to worry about cases like these.

e-kolpakov replied:
@clintonb I can't guarantee there are no "poisonous" data already in staging/production DBs, so at least for initial phase this code should be tolerant to such cases

clintonb replied:
This is the first time we are actually using the fields. Anyone configuring a Site will need to update these values. We should be as strict as possible here. If your site breaks because the values are wrong, that's a good thing. You need to fix your site.

e-kolpakov replied:
@clintonb ok.

clintonb commented on ecommerce/extensions/payment/helpers.py:
```diff
@@ -87,3 +89,15 @@ def sign(message, secret):
+
+
+def get_payment_processors():
+    request = get_current_request()
```
Don't bother with raising an error. Pass the request as an argument and let the caller worry about where that request comes from. Better yet, make this an instance method on `SiteConfiguration`.

clintonb commented on ecommerce/settings/_oscar.py:
```diff
@@ -107,23 +107,25 @@
-        'receipt_path': PAYMENT_PROCESSOR_RECEIPT_PATH,
-        'cancel_path': PAYMENT_PROCESSOR_CANCEL_PATH,
-        'error_path': PAYMENT_PROCESSOR_ERROR_PATH,
+    'edX': {
```
Always compare the lowercase partner code to avoid issues with case sensitivity.

clintonb commented on ecommerce/extensions/payment/management/commands/paypal_profile.py:
```diff
@@ -47,9 +50,33 @@ def handle(self, *args, **options):
+        action = args.pop(0)
+
+        if partner not in settings.PAYMENT_PROCESSOR_CONFIG:
+            raise CommandError("Payment Processor configuration for partner {0} not found".format(partner))
```
Single quotes or double? Pick one is be consistent throughout the file.

clintonb commented on ecommerce/extensions/payment/management/commands/paypal_profile.py:
```diff
@@ -47,9 +50,33 @@ def handle(self, *args, **options):
+        if len(args) < 2:
+            raise CommandError("Required arguments 'config' and 'action' are missing")
+
+        partner = args.pop(0)
```
Lowercase this value.

clintonb commented on ecommerce/extensions/payment/management/commands/paypal_profile.py:
```diff
@@ -47,9 +50,33 @@ def handle(self, *args, **options):
+        partner = args.pop(0)
+        action = args.pop(0)
+
+        if partner not in settings.PAYMENT_PROCESSOR_CONFIG:
```
Combine these two checks: `if self.PAYPAL_CONFIG_KEY not in settings.PAYMENT_PROCESSOR_CONFIG.get(partner, {})`.

clintonb commented on ecommerce/extensions/payment/processors/__init__.py:
```diff
@@ -59,7 +62,11 @@ def configuration(self):
             KeyError: If no settings found for this payment processor
         """
-        return settings.PAYMENT_PROCESSOR_CONFIG[self.NAME]
+        request = get_current_request()
```
I really don't like this pattern of looking for a request. I would much rather the request be a parameter on the methods/constructors.

douglashall replied:
Maybe move the pattern into a custom `get_current_request()` function? The issue with requiring a a request parameter is that you are just pushing the pattern upstream.

clintonb replied:
I'm okay with pushing the pattern upstream, especially given the lack of documentation of the request requirement and exception. I would much rather know that I need a request to instantiate the class than learn about it via an unhandled exception later. I see no place where we are catching `MissingRequestError` outside of tests. This code would be much safer if it were explicit about its need for a request.

e-kolpakov replied:
@clintonb @douglashall Initial implementation passed configuration to payment processors constructor/initializer (i.e. `__init__` - some are just pedantic about it not being an actual constructor :)). However, there was a problem somewhere, so I had to revert to getting settings in base class property from using threadlocal request. I can't recall what was the exact problem (likely it payment processor was created in some place *very* far from where `request` is available) - I'll try again, but I trust my past self on it, so likely it'll have to stay as it is.

e-kolpakov replied:
@clintonb @douglashall yes, the problematic place is `Refund._issue_credit:

     processor = get_processor_class_by_name(source.source_type.name)()

The longest call chain is as follows:

* RefundCreateView.create - request available as parameter
* refund.api.create_refunds
* Refund.create_with_lines
* Refund.approve
* Refund._issue_credit

So, to me using threadlocal request is a lesser evil than passing the request / payment processor configuration through that four calls.

clintonb replied:
Leave it, but document all errors being raised.

clintonb commented on ecommerce/extensions/payment/processors/paypal.py:
```diff
@@ -47,6 +47,14 @@ def __init__(self):
         self.retry_attempts = self.configuration.get('retry_attempts', 1)
 
     @property
```
Use `@cached_property` to avoid instantiating this multiple times.

clintonb commented on ecommerce/extensions/payment/tests/test_processors.py:
```diff
@@ -61,6 +62,18 @@ class PaymentProcessorTestCaseMixin(RefundTestMixin, CourseCatalogTestMixin, Pay
+    def setUpClass(cls):
+        super(PaymentProcessorTestCaseMixin, cls).setUpClass()
+        paypal_configuration = settings.PAYMENT_PROCESSOR_CONFIG['edX'].get('paypal')
+        if paypal_configuration:
```
Why is this necessary given that the payment processor does this already?

e-kolpakov replied:
@clintonb the test itself creates a lot of paypal SDK objects, and they require the api, either statically configured (like this), or passed as constructor parameter (like in `Paypal` class). Without this, the test fails with

```
Traceback (most recent call last):
  File "/work/OpenCraft/edX/analytics-devstack/ecommerce/ecommerce/extensions/payment/tests/test_processors.py", line 631, in test_issue_credit_error
    paypal_refund = paypalrestsdk.Refund({'error': expected_response})
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/paypalrestsdk/resource.py", line 14, in __init__
    self.__dict__['api'] = api or default_api()
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/paypalrestsdk/api.py", line 311, in default
    Refer https://github.com/paypal/rest-api-sdk-python#configuration")
MissingConfig: Required PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET.  
```

clintonb replied:
Please add this reasoning as a comment.

Also, what happens if `paypal_configuration` is empty? If this is required, the line above should use `['paypal']` instead of `get()`.

clintonb commented on ecommerce/extensions/payment/tests/test_processors.py:
```diff
@@ -50,6 +50,7 @@
 log = logging.getLogger(__name__)
 
 
+@ddt.ddt
```
This file has grown a bit too large. My cutoff is around 400 lines. Please break this out, preferably in a separate commit, into multiple files, one per processor.

clintonb commented:
Looks good, but I would like to see the outstanding comments addressed before this is merged.

e-kolpakov commented:
@clintonb I've addressed your comments (except one, which I would like to discuss first) - please take another look.

e-kolpakov commented:
@clintonb still need your approval on this.

@douglashall the contents have changed since your approval (not much, but still) - does your +1 still holds?

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +50,35 @@ class SiteConfiguration(models.Model):
 
+    @property
+    def payment_processors_set(self):
+        return {
```
This should return a `list`. There is no need to return a `dict`, given that you only ever look at the keys.

douglashall replied:
This is a set comprehension. A set will be returned.

clintonb replied:
Ah. Add a docstring, please.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +50,35 @@ class SiteConfiguration(models.Model):
+
+    def _clean_payment_processors(self):
+        if self.payment_processors.isspace():
+            raise ValidationError("Invalid payment processors field: must not only contain whitespace characters")
```
Single quotes.

e-kolpakov replied:
@clintonb ok, but I would like to point out that most of the user-facing strings use double quotes. And "internal" strings (dictionary keys, internal code, etc.) use single quote. And in some places these rules are violated, including this file (i.e. `theme_scss_path` and `payment_processors` help_texts).

clintonb replied:
You have the dubious honor of being the first, and probably, only person's code I review today. That means you get the well-rested reviewer, hence more strict about using one quote style throughout a file/module/codebase.

`lms_url_root` and `payment_processors` are exceptions because they include single quotes in the help text. If you see violations, please fix them. Others might not be as strict as I am, hence the mix of styles.

e-kolpakov replied:
@clintonb sure, but could I have a list of rules to follow on this? It looks like it is as foolows:

1. If the string contains `'` or `"` in body, use other quotes to avoid escaping.
2. If it is a docstring use three double quotes
3. If it is an internal thing (i.e. dict key, `getattr` argument, etc.) - us single quote.
4. Otherwise use double quote.

Is that list correct and complete?

clintonb replied:
The only time you need double quotes are for docstrings and strings containing a single quote that you don't want to escape. That's the de-facto rule that many of us have been following.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +50,35 @@ class SiteConfiguration(models.Model):
+        }
+
+    def _clean_payment_processors(self):
+        if self.payment_processors.isspace():
```
You're really like this `isspace()` case. This can be improved a tad:

```python
def _clean_payment_processors(self):
  value = self.payment_processors.strip()
  if not value:
    raise ValidationError('The payment processors field must contain a non-empty value.')

processor_names = value.split(',')
for name in processor_names:
  try:
    get_processor_class_by_name(name.strip())
  except ProcessorNotFoundError:
    raise ValidationError('[{name}] is not a valid payment processor.' .format(name=name))

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+        ("paypal,cybersource,something_else", {"paypal", "cybersource", "something_else"}),
+    )
+    @ddt.unpack
+    def test_allowed_payment_processors(self, payment_processors, expected_result):
```
1. Give me a docstring for every test explaining what it's testing.
2. This test is more appropriately named `test_payment_processor_field_parsing`.

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+        site_config = SiteConfiguration(payment_processors=payment_processors)
+        self.assertEqual(site_config.payment_processors_set, expected_result)
+
+    @mock.patch("ecommerce.extensions.payment.helpers.get_processor_class_by_name")
```
Nit: You can move this inside the test method.

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+                err.message, "Invalid payment processors field: must not only contain whitespace characters"
+            )
+
+    @mock.patch("ecommerce.extensions.payment.helpers.get_processor_class_by_name")
```
You don't need to patch this.

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+
+            site_config.payment_processors = "irrelevant-get_processor_by_name-is-patched"
+            site_config.clean_fields(exclude={"payment_processors"})
+        except ValidationError as exc:
```
Don't bother catching the exception. Raise exceptions will fail the test.

e-kolpakov replied:
@clintonb this way is slightly semantically better as:

1. It explicitly expresses that test checks for raised/not raised exception (of particular type); without this a comment like "if it raises - test failed" is needed.
2. It expects a particular exception to be not raised, so raising that exception is technically a failed test, not an error.

clintonb replied:
Regardless of the exception raised, an exception is a failure. All you're doing by catching the exception is hiding the stack trace, which would actually help debug the broken test.

e-kolpakov replied:
@clintonb technically, non-catched exception is an error, not a failure; little attention is paid to the difference (as both "fail" the entire suite/build/etc.), but to me "failure" is an expected thing - we expect this to do that and it was not the case, while "error" is - we expect this to do that, but something totally unrelated prevented it from doing so.

Anyway, in order to not engage into a prolonged debate for three lines of code, I'll remove the `try..except` clause.

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+        with self.assertRaises(ValidationError):
+            site_config.clean_fields()
+
+    @mock.patch("ecommerce.extensions.payment.helpers.get_processor_class_by_name")
```
No need to patch.

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+        'ecommerce.extensions.payment.tests.processors.AnotherDummyProcessor',
+    ])
+    @ddt.data(
+        ([], []),
```
This is not a valid case.

e-kolpakov replied:
@clintonb it is valid for the method under test; `payment_processor` field "non-emptiness" is enforced by validation code and `blank=False`, `null=False` in field declaration, which will make this test depend on validation/field configuration, i.e. make it more brittle.

clintonb commented on ecommerce/core/tests/test_models.py:
```diff
@@ -45,3 +54,110 @@ class BusinessClientTests(TestCase):
+    @ddt.data(
+        ([], []),
+        ([DummyProcessor], [DummyProcessor]),
+        ([DummyProcessor, AnotherDummyProcessor], [DummyProcessor]),
```
This seems wrong. If only a subset of processors exists, the others should not be in the list, and an error should be raised.

e-kolpakov replied:
@clintonb there are at least one somewhat valid use case for this:

* Initially both Dummy and AnotherDummy processors was enabled in settings
* SiteConfiguration with both Dummy and AnotherDummy processor was created
* AnotherDummy was later removed from settings.

In two words, list of "available" processors can change after a model is created, and there's no code so far that could protect from such change.

Also, this test tests the method, not a model itself, so it can't rely on the fact that validation will fail a record with unknown processor.

clintonb replied:
If there is concern about settings changing, and breaking the model, we should have startup validation. The app doesn't start unless every `Site` has access to its listed payment processors. At the very least, if you don't want to go the validation route, you should log a warning whenever the `Site` expects a payment processor that is not configured in settings.

I would much rather have startup validation given that we have different settings across environments. Things might be fine in stage, where we run acceptance tests. However, we don't test production after a release. We would have to wait for error alerts or support calls. If something is going to break, it should break as early in the process as possible.

e-kolpakov replied:
@clintonb ok, I'll see if I could add validating site configurations at startup.

clintonb commented on ecommerce/extensions/api/v2/tests/views/test_payments.py:
```diff
@@ -16,8 +17,13 @@ class PaymentProcessorListViewTests(TestCase):
+        self.toggle_payment_processor(AnotherDummyProcessor.NAME, True)
+
+        site_config, __ = SiteConfiguration.objects.get_or_create(site__id=1)
+        site_config.payment_processors = ",".join([DummyProcessor.NAME, AnotherDummyProcessor.NAME])
```
Use `update_or_create` and pass this value in the defaults dict.

clintonb replied:
Alternatively, just use `self.site.siteconfiguration`.

e-kolpakov replied:
@clintonb There's a problem with `update_or_create`, sine it requires passing partner_id, site_id and so on, and my best guesses was not right (site_id=1 and partner_id=1 caused DB integrity violation), so it looks like `get_or_create` is more robust.

`self.site.site_configuration` raises `AttributeError`

clintonb replied:
`self.site.siteconfiguration`

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +50,35 @@ class SiteConfiguration(models.Model):
+            except ProcessorNotFoundError as exc:
+                raise ValidationError(exc.message)
+
+    def get_payment_processors(self):
```
Add a docstring explaining what this returns.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +50,35 @@ class SiteConfiguration(models.Model):
+            if processor.NAME in self.payment_processors_set and processor.is_enabled()
+        ]
+
+    def clean_fields(self, exclude=None):
```
Docstring.

e-kolpakov replied:
@clintonb self-explanatory, but ok, I'll add it.

clintonb commented on ecommerce/extensions/basket/tests/test_views.py:
```diff
@@ -122,6 +123,10 @@ def setUp(self):
         self.user = self.create_user()
         self.client.login(username=self.user.username, password=self.password)
         self.course = CourseFactory(name='BasketSummaryTest')
+        self.site_configuration = SiteConfiguration.objects.get(site__id=1)
```
`self.site_configuration = self.site.site_configuration`

e-kolpakov replied:
@clintonb does not work like that:

```
Traceback (most recent call last):
  File "/work/OpenCraft/edX/analytics-devstack/ecommerce/ecommerce/extensions/basket/tests/test_views.py", line 127, in setUp
    site_configuration = self.site.site_configuration
AttributeError: 'Site' object has no attribute 'site_configuration'
```

clintonb replied:
My mistake. This is not a foreign key relationship. In that case, minor improvement:

```python
self.site_configuration, __ = SiteConfiguration.objects.update_or_create(site__id=self.site.id, defaults={'payment_processors': DummyProcessor.NAME})

clintonb replied:
`self.site.siteconfiguration`

clintonb commented on ecommerce/extensions/payment/tests/processors/base.py:
```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
```
Rename to mixins.py.

clintonb commented:
Here's what needs to be done to earn my approval:
- [ ] Remaining comments addressed
- [ ] Some form of validation added to ensure `Site`s are not configured to use payment processors that are not in settings. The last thing we want is to display a checkout form to the user with no ability to pay.

douglashall commented:
@e-kolpakov This still looks good to me. Thanks for all your work on this.

e-kolpakov commented on ecommerce/wsgi.py:
```diff
@@ -32,3 +31,11 @@
 # application = HelloWorldApplication(application)
+
+
+def application_ready():
```
@clintonb here's the motivation for it:

1. In order to validate models field values we need to fetch them from DB.
2. We can't do that in AppConfig, as [Django docs](https://docs.djangoproject.com/en/1.9/ref/applications/#django.apps.AppConfig.ready) explicitly tells it's a bad idea (and I've tried that anyway and it couldn't even run the tests)
3. We can't do that in `models.py`, as SiteConfiguration references model from other app and it is not ready (and likely that's just the first problem on the way; most likely it won't work anyway, since `syncdb` needs models to be ready to create DB, and we're trying to access DB before models module is processed)
4. I've considered using middleware for this, but it would be executed on every request, which means +N SQL queries for evry request (N = SiteConfiguration). It could be solved by some global flag, but it felt more and more hackish, so I have decided to not follow that path.

And that's how I ended with this - it is executed once, fails startup (raises ValidateException, does not respond to HTTP requests at all) and covered by tests to the maximum possible extent.

If you've got better ideas for how to achieve this - let me know.

clintonb replied:
I would prefer to see this in `AppConfig.ready()`, despite Django's warning. We are definitely not modifying any data, and the handful of instances we'd be iterating over are not enough to drastically slow down execution of management commands.

You'll most likely have to either duplicate code, or move it to a utility function/module, to avoid importing `models.py` at startup.

e-kolpakov replied:
@clintonb I've tried putting it into `AppConfig.ready` and it didn't work: AppConfig.ready is executed before an app can sync models to DB, so even reading form DB is not possible. In particular:

```
# core.config.py
from django.apps import AppConfig

class CoreAppConfig(AppConfig):
    name = 'ecommerce.core'
    verbose_name = 'Core'

    def ready(self):
        super(CoreAppConfig, self).ready()

        # Ensures that the initialized Celery app is loaded when Django starts.
        # Allows Celery tasks to bind themselves to an initialized instance of the Celery library.
        from ecommerce import celery_app  # pylint: disable=unused-variable

        from .models import validate_configuration
        validate_configuration()
```

Crashed with

```
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 328, in execute
    django.setup()
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/apps/registry.py", line 115, in populate
    app_config.ready()
  File "/work/OpenCraft/edX/analytics-devstack/ecommerce/ecommerce/core/config.py", line 16, in ready
    validate_configuration()
  File "/work/OpenCraft/edX/analytics-devstack/ecommerce/ecommerce/core/models.py", line 152, in validate_configuration
    for config in SiteConfiguration.objects.all():
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/query.py", line 162, in __iter__
    self._fetch_all()
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/query.py", line 965, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/query.py", line 238, in iterator
    results = compiler.execute_sql()
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 840, in execute_sql
    cursor.execute(sql, params)
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/john/.virtualenvs/ecommerce/local/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py", line 318, in execute
    return Database.Cursor.execute(self, query, params)
OperationalError: no such table: core_siteconfiguration
```

clintonb replied:
Catch, and log, the `OperationalError`. Add a comment stating why it's safe to ignore for tests.

smarnach replied:
When running `./manage.py migrate` with a completely empty DB, I got a `ProgrammingError` here rather than an `OperationalError`:
```
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 328, in execute
    django.setup()
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/apps/registry.py", line 115, in populate
    app_config.ready()
  File "/edx/app/ecommerce/ecommerce/ecommerce/core/config.py", line 30, in ready
    validate_configuration()
  File "/edx/app/ecommerce/ecommerce/ecommerce/core/models.py", line 158, in validate_configuration
    for config in SiteConfiguration.objects.all():
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/query.py", line 162, in __iter__
    self._fetch_all()
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/query.py", line 965, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/query.py", line 238, in iterator
    results = compiler.execute_sql()
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 840, in execute_sql
    cursor.execute(sql, params)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 124, in execute
    return self.cursor.execute(query, args)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/edx/app/ecommerce/venvs/ecommerce/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
django.db.utils.ProgrammingError: (1146, "Table 'ecommerce.core_siteconfiguration' doesn't exist")
```
I got there by dropping and recreating the `ecommerce` database manually, but I suspect the same thing would happen on a fresh install.  I've got no idea why a missing table results in different exceptions being raised, but simply adding `ProgrammingError` to the list of ignored exceptions fixed the problem for me.

bradenmacdonald replied:
@smarnach Yep, the ECOM team discovered that and we have https://openedx.atlassian.net/browse/WL-390 planned to fix it.

e-kolpakov commented on Makefile:
```diff
@@ -78,6 +79,12 @@ static:
+
+fast_diff_coverage:
+	coverage xml
+	diff-cover coverage.xml --compare-branch=$(DIFF_COVER_BASE_BRANCH)
```
@clintonb allows checking diff coverage locally and not wait for push, CI build and codecov processing.

clintonb replied:
Diff coverage isn't necessary if you ensure your code has 100% coverage.

e-kolpakov replied:
@clintonb this is a tool to help ensure my code have 100% coverage.

clintonb replied:
Okay.

e-kolpakov commented:
@clintonb addressed remaining comments (including validating models on startup), please take another look.

P.S. sorry if I'm getting bothersome - merge deadline is close (I've just added it to description, sorry for not communicating it earlier), so I would like to speed up the process (in terms of results, not the effort put)

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +55,63 @@ class SiteConfiguration(models.Model):
+        missing_processor_configurations = self.payment_processors_set - all_processor_names
+        if missing_processor_configurations:
+            processor_config_repr = ", ".join(missing_processor_configurations)
+            log.warning(
```
If you're using this for startup validation, logging a warning won't be good enough. You need to raise an error to actually stop execution.

e-kolpakov replied:
@clintonb no, validation is performed in:

* wsgi.py -> application_ready
* core.models -> validate_configuration
* core.models.SiteConfiguration -> clean_fields

clintonb replied:
Ah. Yes. My mistake.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -45,6 +55,63 @@ class SiteConfiguration(models.Model):
+            try:
+                get_processor_class_by_name(name.strip())
+            except ProcessorNotFoundError as exc:
+                raise ValidationError(exc.message)
```
Add a call to `log.exception()` as well, so we don't lose the stack trace.

mattdrayer commented:
@saleem-latif, just FYI on the work being done in this branch

e-kolpakov commented:
@clintonb anything else needs to be done here?

clintonb commented on ecommerce/core/config.py:
```diff
@@ -1,13 +1,32 @@
+        # just fine with this suppression - most likely test code hit this line before the DB is populated
+        try:
+            # Here we're are consciously violating Django's warning about not interacting with DB in AppConfig.ready
+            # We now what we're doing, have considered a couple of other approaches and discussed it in great length:
```
s/now/know.

clintonb commented:
:+1: pending last comment addressed, commits squashed

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! I've created [OSPR-1694](https://openedx.atlassian.net/browse/OSPR-1694) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/ecommerce/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

e-kolpakov commented:
@edx-webhook That's so kind of you, thank you! Please note this PR was merged like a year ago though. :)

gsong commented:
@e-kolpakov Yeah, we're investigating some bugs with the bot.
